File: jbolo/src/com/jbolo/gui/Sound.java
/*
 * $Id$
 *
 * Copyright (c) 1998-2008 John Morrison.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 */


/*
 * Sound.java
 *
 * Created on December 30, 2003, 9:43 PM
 */

package com.jbolo.gui;

import java.applet.Applet;
import java.applet.AudioClip;
import java.net.URL;
import javax.sound.sampled.*;
import java.io.*;
import com.jbolo.bolo.SoundDist;



/** Provides sound output
 * @author John Morrison
 */
public class Sound implements FrontEndSound {
    
    /** The base path for all sounds to be loaded from */
    private String baseUrlPath;
    /** Are we playing any sounds at all */
    private boolean isPlayable;
    /** Is sounds enabled in the menu */
    private boolean isEnabledInMenu;
    /** Sound Effect Audio Clip */
    private Clip clipBigExplosionFar;
    /** Sound Effect Audio Clip */
    private Clip clipBigExplosionNear;
    /** Sound Effect Audio Clip */
    private Clip clipBubbles;
    /** Sound Effect Audio Clip */
    private Clip clipFarmingTreeFar;
    /** Sound Effect Audio Clip */
    private Clip clipFarmingTreeNear;
    /** Sound Effect Audio Clip */
    private Clip clipHitTankFar;
    /** Sound Effect Audio Clip */
    private Clip clipHitTankNear;
    /** Sound Effect Audio Clip */
    private Clip clipHitTankSelf;
    /** Sound Effect Audio Clip */
    private Clip clipManBuildingFar;
    /** Sound Effect Audio Clip */
    private Clip clipManBuildingNear;
    /** Sound Effect Audio Clip */
    private Clip clipManDyingFar;
    /** Sound Effect Audio Clip */
    private Clip clipManDyingNear;
    /** Sound Effect Audio Clip */
    private Clip clipManLayingMineNear;
    /** Sound Effect Audio Clip */
    private Clip clipMineExplosionFar;
    /** Sound Effect Audio Clip */
    private Clip clipMineExplosionNear;
    /** Sound Effect Audio Clip */
    private Clip clipShootFar;
    /** Sound Effect Audio Clip */
    private Clip clipShootNear;
    /** Sound Effect Audio Clip */
    private Clip clipShootSelf;
    /** Sound Effect Audio Clip */
    private Clip clipShotBuildingFar;
    /** Sound Effect Audio Clip */
    private Clip clipShotBuildingNear;
    /** Sound Effect Audio Clip */
    private Clip clipShotTreeFar;
    /** Sound Effect Audio Clip */
    private Clip clipShotTreeNear;
    /** Sound Effect Audio Clip */
    private Clip clipTankSinkingFar;
    /** Sound Effect Audio Clip */
    private Clip clipTankSinkingNear;
    
    private SoundEffect test;
    
    int test2;
    
    /** Creates a new instance of Sound */
    public Sound() {
	isPlayable = false;
	isEnabledInMenu = false;
	baseUrlPath = "file:///c:/jbolo/sounds/";
	//	clip = Applet.newAudioClip(url);
	//	clip.loop();
	//	    clip = applet.getAudioClip(url);
	//	    clip.play();
    }
    
    /** Gets if sounds are enabled
     * @return If sounds are enabled
     */
    public boolean canPlay() {
	return isPlayable;
    }
    
    /**
     * Gets if sounds are enabled in the menu
     * @return If the sound is enabled
     */
    public boolean isEnabled() {
	return isEnabledInMenu;
    }
    
    /**
     * Sets whether sounds are enabled in the menus
     * @param value Value to set
     */
    public void setEnabled(boolean value) {
	isEnabledInMenu = value;
    }
    
    /**
     * Sets up and loads all sounds
     * @return If all were loaded successfully or not 
     */
    public boolean setup() {
	isPlayable = true;
	System.err.print("Loading sounds...");
	try {
        test = new SoundEffect(new URL(this.baseUrlPath + "shooting_self.wav"));

            clipBigExplosionFar = loadSound("big_explosion_far.wav");
	    clipBigExplosionNear = loadSound("big_explosion_near.wav");
	    clipBubbles = loadSound("bubbles.wav");
	    clipFarmingTreeFar = loadSound("farming_tree_far.wav");
	    clipFarmingTreeNear = loadSound("farming_tree_near.wav");
	    clipHitTankFar = loadSound("hit_tank_far.wav");
	    clipHitTankNear = loadSound("hit_tank_near.wav");
	    clipHitTankSelf = loadSound("hit_tank_self.wav");
	    clipManBuildingFar = loadSound("man_building_far.wav");
	    clipManBuildingNear = loadSound("man_building_near.wav");
	    clipManDyingFar = loadSound("man_dying_far.wav");
	    clipManDyingNear = loadSound("man_dying_near.wav");
	    clipManLayingMineNear = loadSound("man_lay_mine_near.wav");
	    clipMineExplosionFar = loadSound("mine_explosion_far.wav");
	    clipMineExplosionNear = loadSound("mine_explosion_near.wav");
	    clipShootFar = loadSound("shooting_far.wav");
	    clipShootNear = loadSound("shooting_near.wav");
	    clipShootSelf = loadSound("shooting_self.wav");
	    clipShotBuildingFar = loadSound("shot_building_far.wav");
	    clipShotBuildingNear = loadSound("shot_building_near.wav");
	    clipShotTreeFar = loadSound("shot_tree_far.wav");
	    clipShotTreeNear = loadSound("shot_tree_near.wav");
	    clipTankSinkingFar = loadSound("tank_sinking_near.wav");
	    clipTankSinkingNear = loadSound("tank_sinking_far.wav");
//	    test = loadSound2("shooting_self.wav");
	} catch (Exception e) {
	    e.printStackTrace();
	    isPlayable = false;
	}
	isEnabledInMenu = isPlayable;
	System.err.println("Done");
	return isPlayable;
    }
    
    /** Loads a sound effect
     * @return Clip containing the loaded sound
     * @throw Exception on load error
     * @param fileName Filename of the sound to load
     * @throws Exception On load error
     */
    private Clip loadSound(String fileName) throws Exception {
	Clip returnValue; // Value to return
        Object currentSound;
        
        System.err.print(".");
        
        
        currentSound = AudioSystem.getAudioInputStream(new URL(this.baseUrlPath + fileName));

        AudioInputStream stream = (AudioInputStream) currentSound;
        AudioFormat format = stream.getFormat();

        /**
         * we can't yet open the device for ALAW/ULAW playback,
         * convert ALAW/ULAW to PCM
         */
        if ((format.getEncoding() == AudioFormat.Encoding.ULAW) ||
            (format.getEncoding() == AudioFormat.Encoding.ALAW)) 
        {
            AudioFormat tmp = new AudioFormat(
                                      AudioFormat.Encoding.PCM_SIGNED, 
                                      format.getSampleRate(),
                                      format.getSampleSizeInBits() * 2,
                                      format.getChannels(),
                                      format.getFrameSize() * 2,
                                      format.getFrameRate(),
                                      true);
            stream = AudioSystem.getAudioInputStream(tmp, stream);
            format = tmp;
        }
        DataLine.Info info = new DataLine.Info(
                                  Clip.class, 
                                  stream.getFormat(), 
                                  ((int) stream.getFrameLength() *
                                      format.getFrameSize()));

        Clip clip = (Clip) AudioSystem.getLine(info);
        clip.open(stream);
        
        return clip;
    
    
    }
    
    private SoundEffect loadSound2(String fileName) throws Exception {
    	return new SoundEffect(new URL(baseUrlPath + fileName));
    }
    
    /**
     * Plays a sound effect of type effectType
     * @param effectType The sound effect type to play
     */
    public void fePlaySound(int effectType) {
	if (isPlayable == true && isEnabledInMenu == true) {
/*	    switch (effectType) {
		case SoundDist.BIG_EXPLOSION_FAR:
		    clipBigExplosionFar.play();
		    break;
		case SoundDist.BIG_EXPLOSION_NEAR:
		    clipBigExplosionNear.play();
		    break;
		case SoundDist.BUBBLES:
		    clipBubbles.play();
		    break;
		case SoundDist.FARMING_TREE_FAR:
		    clipFarmingTreeFar.play();
		    break;
		case SoundDist.FARMING_TREE_NEAR:
		    clipFarmingTreeNear.play();
		    break;
		case SoundDist.HIT_TANK_FAR:
		    clipHitTankFar.play();
		    break;
		case SoundDist.HIT_TANK_NEAR:
		    
		    clipHitTankNear.play();
		    break;
		case SoundDist.HIT_TANK_SELF:
		    clipHitTankSelf.play();
		    break;
		case SoundDist.MAN_BUILDING_FAR:
		    clipManBuildingFar.play();
		    break;
		case SoundDist.MAN_BUILDING_NEAR:
		    clipManBuildingNear.play();
		    break;
		case SoundDist.MAN_DYING_FAR:
		    clipManDyingFar.play();
		    break;
		case SoundDist.MAN_DYING_NEAR:
		    clipManDyingNear.play();
		    break;
		case SoundDist.MAN_LAYING_MINE_NEAR:
		    clipManLayingMineNear.play();
		    break;
		case SoundDist.MINE_EXPLOSION_FAR:
		    clipMineExplosionFar.play();
		    break;
		case SoundDist.MINE_EXPLOSION_NEAR:
		    clipMineExplosionNear.play();
		    break;
		case SoundDist.SHOOT_FAR:
		    clipShootFar.play();
		    break;
		case SoundDist.SHOOT_NEAR:
		    clipShootNear.play();
		    break;
		case SoundDist.SHOOT_SELF:
		    clipShootSelf.play();
		    break;
		case SoundDist.SHOT_BUILDING_FAR:
		    clipShotBuildingFar.play();
		    break;
		case SoundDist.SHOT_BUILDING_NEAR:
		    clipShotBuildingNear.play();
		    break;
		case SoundDist.SHOT_TREE_FAR:
		    clipShotTreeFar.play();
		    break;
		case SoundDist.SHOT_TREE_NEAR:
		    clipShotTreeNear.play();
		    break;
		case SoundDist.TANK_SINK_FAR:
		    clipTankSinkingFar.play();
		    break;
		case SoundDist.TANK_SINK_NEAR:
		    clipTankSinkingNear.play();
		default:
		    System.err.println("Sound error: " + effectType);
		    break;
		    
	    } */
	}
    }
}

File: winbolo/src/gui/linux/sound.c
/*
 * $Id$
 *
 * Copyright (c) 1998-2008 John Morrison.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 */


/*********************************************************
*Name:          Sound
*Filename:      sound.c
*Author:        John Morrison
*Creation Date: 26/12/98
*Last Modified:  13/6/00
*Purpose:
*  System Specific Sound Playing routines 
*  (Uses Direct Sound)
*********************************************************/

#include <stdio.h>
#include <unistd.h>
#include <glib.h>
#include "SDL.h"
#include "../../bolo/global.h"
#include "../../bolo/backend.h"
#include "bolosounds.h"
#include "messagebox.h"
#include "../lang.h"
#include "../linresource.h"

//#include "..\dsutil.h"
//#include "..\lang.h"
//#include "..\sound.h"

extern gchar *applicationPath;

#define MAX_CHANNELS 8

Mix_Chunk *lpDSPrimary = NULL;
Mix_Chunk *lpDSBigExplosionFar = NULL;
Mix_Chunk *lpDSBigExplosionNear = NULL;
Mix_Chunk *lpDSBubbles = NULL;
Mix_Chunk *lpDSFarmingTreeFar = NULL;
Mix_Chunk *lpDSFarmingTreeNear = NULL;
Mix_Chunk *lpDSHitTankFar = NULL;
Mix_Chunk *lpDSHitTankNear = NULL;
Mix_Chunk *lpDSHitTankSelf = NULL;
Mix_Chunk *lpDSManBuildingFar = NULL;
Mix_Chunk *lpDSManBuildingNear = NULL;
Mix_Chunk *lpDSManDyingFar = NULL;
Mix_Chunk *lpDSManDyingNear = NULL;
Mix_Chunk *lpDSManLayingMineNear = NULL;
Mix_Chunk *lpDSMineExplosionFar = NULL;
Mix_Chunk *lpDSMineExplosionNear = NULL;
Mix_Chunk *lpDSShootFar = NULL;
Mix_Chunk *lpDSShootNear = NULL;
Mix_Chunk *lpDSShootSelf = NULL;
Mix_Chunk *lpDSShotBuildingFar = NULL;
Mix_Chunk *lpDSShotBuildingNear = NULL;
Mix_Chunk *lpDSShotTreeFar = NULL;
Mix_Chunk *lpDSShotTreeNear = NULL;
Mix_Chunk *lpDSTankSinkingFar = NULL;
Mix_Chunk *lpDSTankSinkingNear = NULL;

bool isPlayable;


void windowDisableSound();

Mix_Chunk *soundLoad(char *tempFile, FILE *fp, BYTE *buff, int len) {
  Mix_Chunk *returnValue;
  FILE *out;
  
  fread(buff, len, 1, fp);
  out = fopen(tempFile, "wb");
  fwrite(buff, len, 1, out);
  fclose(out); 
  returnValue = Mix_LoadWAV(tempFile);
  return returnValue;
}

/*********************************************************
*NAME:          soundSetup
*AUTHOR:        John Morrison
*CREATION DATE: 26/10/98
*LAST MODIFIED:  13/6/00
*PURPOSE:
*  Sets up sound systems, direct sound structures etc.
*  Returns whether the operation was successful or not
*
*ARGUMENTS:
* appInst - Handle to the application (Required to 
*           load bitmaps from resources)
* appWnd  - Main Window Handle (Required for clipper)
*********************************************************/
bool soundSetup() {
  bool returnValue;       /* Value to return */
  bool dsLoadOK;
  gchar *tmpPath;
  char fileName[FILENAME_MAX];
  char inFile[FILENAME_MAX];
  char *tf;
  FILE *in;
  BYTE *buff;

  buff = malloc(1024 * 100);
  /* Get tmp file */
  strcpy(fileName, g_get_tmp_dir());
  tf = tempnam(fileName, "lb");
  strcpy(fileName, tf);

  /* Get in path */
  strcpy(inFile, applicationPath);
  strcat(inFile, "/");
  strcat(inFile, "bolosounds.bsd");


  returnValue = TRUE;
  dsLoadOK = TRUE;
  isPlayable = TRUE;

  in = fopen(inFile, "rb");
  if (!in) {
//    windowDisableSound();
    isPlayable = FALSE;
  //  MessageBox(langGetText(STR_SOUNDERR_FILENOTFOUND), DIALOG_BOX_TITLE);
  }
	
  
  if (returnValue == TRUE && isPlayable == TRUE) {
    if (Mix_OpenAudio(22050, AUDIO_S16, 2, 512) < 0) {
      MessageBox("Couldn't start sound. Check to make sure\nthe device is not busy", DIALOG_BOX_TITLE);
      isPlayable = FALSE;
      returnValue = TRUE;
    }
  }
  
  if (returnValue == TRUE && isPlayable == TRUE) {
  /* Load Sound Files In */
    lpDSBigExplosionFar = soundLoad(fileName, in, buff, 31332);
    lpDSBigExplosionNear = soundLoad(fileName, in, buff, 31332);
    lpDSBubbles = soundLoad(fileName, in, buff, 7248);
    lpDSFarmingTreeFar = soundLoad(fileName, in, buff, 14480);
    lpDSFarmingTreeNear = soundLoad(fileName, in, buff, 14480);
    lpDSHitTankFar = soundLoad(fileName, in, buff, 6090);
    lpDSHitTankNear =soundLoad(fileName, in, buff, 6048);
    lpDSHitTankSelf = soundLoad(fileName, in, buff, 15632);
    lpDSManBuildingFar = soundLoad(fileName, in, buff, 13534);
    lpDSManBuildingNear = soundLoad(fileName, in, buff, 13534);
    lpDSManDyingFar = soundLoad(fileName, in, buff, 29424);
    lpDSManDyingNear = soundLoad(fileName, in, buff, 29424);
    lpDSManLayingMineNear = soundLoad(fileName, in, buff, 10672);
    lpDSMineExplosionFar = soundLoad(fileName, in, buff, 16336);
    lpDSMineExplosionNear = soundLoad(fileName, in, buff, 16336);
    lpDSShootFar = soundLoad(fileName, in, buff, 9512);
    lpDSShootNear = soundLoad(fileName, in, buff, 14556); 
    lpDSShootSelf = soundLoad(fileName, in, buff, 44464);
    lpDSShotBuildingFar = soundLoad(fileName, in, buff, 13936);
    lpDSShotBuildingNear = soundLoad(fileName, in, buff, 13936);
    lpDSShotTreeFar = soundLoad(fileName, in, buff, 18544);
    lpDSShotTreeNear = soundLoad(fileName, in, buff, 18544);
    lpDSTankSinkingFar = soundLoad(fileName, in, buff, 72252);
    lpDSTankSinkingNear = soundLoad(fileName, in, buff, 72252);
    if (lpDSBigExplosionFar == NULL || lpDSBigExplosionNear == NULL || lpDSBubbles == NULL || lpDSFarmingTreeFar == NULL || lpDSFarmingTreeNear == NULL || lpDSHitTankFar == NULL || lpDSHitTankNear == NULL || lpDSHitTankSelf == NULL || lpDSManBuildingFar == NULL || lpDSManBuildingNear == NULL || lpDSManDyingFar == NULL || lpDSManDyingNear == NULL || lpDSManLayingMineNear == NULL || lpDSMineExplosionFar == NULL || lpDSMineExplosionNear == NULL || lpDSShootFar == NULL || lpDSShootNear == NULL || lpDSShootSelf == NULL || lpDSShotBuildingFar == NULL || lpDSShotBuildingNear == NULL || lpDSShotTreeFar == NULL || lpDSShotTreeNear == NULL || lpDSTankSinkingFar == NULL || lpDSTankSinkingNear == NULL) {
      returnValue = FALSE;
      MessageBox(langGetText(STR_SOUNDERR_LOADSOUNDFAILED), DIALOG_BOX_TITLE);
    }

  }

  if (dsLoadOK == FALSE) {
    returnValue = TRUE;
  }
  if (returnValue == FALSE) {
    isPlayable = FALSE;
//    windowDisableSound();
  }
  unlink(fileName);
  free(buff);
  return returnValue;
}

/*********************************************************
*NAME:          soundCleanup
*AUTHOR:        John Morrison
*CREATION DATE: 26/12/98
*LAST MODIFIED: 28/12/98
*PURPOSE:
*  Destroys and cleans up sound systems, direct sound
*  structures etc.
*
*ARGUMENTS:
*
*********************************************************/
void soundCleanup(void) {
  
  /* Unload the Direct Sound Buffers and Object */
  if (lpDSBigExplosionFar != NULL) {
    Mix_FreeChunk(lpDSBigExplosionFar);
    lpDSBigExplosionFar = NULL;
  }
  if (lpDSBigExplosionNear != NULL) {
    Mix_FreeChunk(lpDSBigExplosionNear);
    lpDSBigExplosionNear = NULL;
  }
  if (lpDSBubbles != NULL) {
    Mix_FreeChunk(lpDSBubbles);
    lpDSBubbles = NULL;
  }
  if (lpDSFarmingTreeFar != NULL) {
    Mix_FreeChunk(lpDSFarmingTreeFar);
    lpDSFarmingTreeFar = NULL;
  }
  if (lpDSFarmingTreeNear != NULL) {
    Mix_FreeChunk(lpDSFarmingTreeNear);
    lpDSFarmingTreeNear = NULL;
  }
  if (lpDSHitTankFar != NULL) {
    Mix_FreeChunk(lpDSHitTankFar);
    lpDSHitTankFar = NULL;
  }
  if (lpDSHitTankNear != NULL) {
    Mix_FreeChunk(lpDSHitTankNear);
    lpDSHitTankNear = NULL;
  }
  if (lpDSHitTankSelf != NULL) {
    Mix_FreeChunk(lpDSHitTankSelf);
    lpDSHitTankSelf = NULL;
  }
  if (lpDSManBuildingFar != NULL) {
    Mix_FreeChunk(lpDSManBuildingFar);
    lpDSManBuildingFar = NULL;
  }
  if (lpDSManBuildingNear != NULL) {
    Mix_FreeChunk(lpDSManBuildingNear);
    lpDSManBuildingNear = NULL;
  }
  if (lpDSManDyingFar != NULL) {
    Mix_FreeChunk(lpDSManDyingFar);
    lpDSManDyingFar = NULL;
  }
  if (lpDSManDyingNear != NULL) {
    Mix_FreeChunk(lpDSManDyingNear);
    lpDSManDyingNear = NULL;
  }

  if (lpDSManLayingMineNear != NULL) {
    Mix_FreeChunk(lpDSManLayingMineNear);
    lpDSManLayingMineNear = NULL;
  }
  if (lpDSMineExplosionFar != NULL) {
    Mix_FreeChunk(lpDSMineExplosionFar);
    lpDSMineExplosionFar = NULL;
  }
  if (lpDSMineExplosionNear != NULL) {
    Mix_FreeChunk(lpDSMineExplosionNear);
    lpDSMineExplosionNear = NULL;
  }
  if (lpDSShotBuildingFar != NULL) {
    Mix_FreeChunk(lpDSShotBuildingFar);
    lpDSShotBuildingFar = NULL;
  }
  if (lpDSShotBuildingNear != NULL) {
    Mix_FreeChunk(lpDSShotBuildingNear);
    lpDSShotBuildingNear = NULL;
  }
  if (lpDSShotTreeFar != NULL) {
    Mix_FreeChunk(lpDSShotTreeFar);
    lpDSShotTreeFar = NULL;
  }
  if (lpDSShotTreeNear != NULL) {
    Mix_FreeChunk(lpDSShotTreeNear);
    lpDSShotTreeNear = NULL;
  }
  if (lpDSTankSinkingFar != NULL) {
    Mix_FreeChunk(lpDSTankSinkingFar);
    lpDSTankSinkingFar = NULL;
  }
  if (lpDSTankSinkingNear != NULL) {
    Mix_FreeChunk(lpDSTankSinkingNear);
    lpDSTankSinkingNear = NULL;
  }
  if (lpDSShootFar != NULL) {
    Mix_FreeChunk(lpDSShootFar);
    lpDSShootFar = NULL;
  }  
  if (lpDSShootNear != NULL) {
    Mix_FreeChunk(lpDSShootNear);
    lpDSShootNear = NULL;
  }
  if (lpDSShootSelf != NULL) {
    Mix_FreeChunk(lpDSShootSelf);
    lpDSShootSelf = NULL;
  }

  if (isPlayable == TRUE) {
    Mix_CloseAudio();
    isPlayable = FALSE;
  }
}


/*********************************************************
*NAME:          soundPlaySound
*AUTHOR:        John Morrison
*CREATION DATE: 28/12/98
*LAST MODIFIED: 28/12/98
*PURPOSE:
*  Plays the specified Sound Buffer
*
*ARGUMENTS:
*  value       - Direct Sound Buffer to load into
*  lpzFileName - File name to load if buffer is lost
*********************************************************/
void soundPlaySound(Mix_Chunk *value, char *name, int channel) {
  static int useChannel = 0;
  if (channel == -1) {
    Mix_PlayChannel(useChannel, value, 0);
    useChannel++;
    if (useChannel == 4 /* MAX_CHANNELS */) {
      useChannel = 0;
    }
  } else {
    Mix_PlayChannel(channel, value, 0);
  }
}

/*********************************************************
*NAME:          soundLoadWave
*AUTHOR:        John Morrison
*CREATION DATE: 28/12/98
*LAST MODIFIED: 28/12/98
*PURPOSE:
*  Plays the correct Sound file
*
*ARGUMENTS:
*  value       - The sound file number to play
*********************************************************/
void soundPlayEffect(sndEffects value) {  
  switch (value) {
  case shootSelf:
    soundPlaySound(lpDSShootSelf, IDW_SHOOTING_SELF, -1);
    break;
  case shootNear:
    soundPlaySound(lpDSShootNear, IDW_SHOOTING_NEAR, -1);
    break;
  case shotTreeNear:
    soundPlaySound(lpDSShotTreeNear, IDW_SHOT_TREE_NEAR, -1);
    break;
  case shotTreeFar:
    soundPlaySound(lpDSShotTreeFar, IDW_SHOT_TREE_FAR, -1);
    break;
  case shotBuildingNear:
    soundPlaySound(lpDSShotBuildingNear, IDW_SHOT_BUILDING_NEAR, -1);
    break;
  case shotBuildingFar:
    soundPlaySound(lpDSShotBuildingFar, IDW_SHOT_BUILDING_FAR, -1);
    break;
  case hitTankFar:
    soundPlaySound(lpDSHitTankFar, IDW_HIT_TANK_FAR, -1);
    break;
  case hitTankNear:
    soundPlaySound(lpDSHitTankNear, IDW_HIT_TANK_NEAR, -1 );
    break;
  case hitTankSelf:
    soundPlaySound(lpDSHitTankSelf, IDW_HIT_TANK_SELF, -1);
    break;
  case bubbles:
    soundPlaySound(lpDSBubbles, IDW_BUBBLES, -1);
    break;
  case tankSinkNear:
    soundPlaySound(lpDSTankSinkingNear, IDW_TANK_SINKING_NEAR, -1);
    break;
  case tankSinkFar:
    soundPlaySound(lpDSTankSinkingFar, IDW_TANK_SINKING_FAR, -1);
    break;
  case bigExplosionNear:
    soundPlaySound(lpDSBigExplosionNear, IDW_BIG_EXPLOSION_NEAR, -1);
    break;
  case bigExplosionFar:
    soundPlaySound(lpDSBigExplosionFar, IDW_BIG_EXPLOSION_FAR, -1);
    break;
  case farmingTreeNear:
    soundPlaySound(lpDSFarmingTreeNear, IDW_FARMING_TREE_NEAR, -1);
    break;
  case farmingTreeFar:
    soundPlaySound(lpDSFarmingTreeFar, IDW_FARMING_TREE_FAR, -1);
    break;
  case manBuildingNear:
    soundPlaySound(lpDSManBuildingNear, IDW_MAN_BUILDING_NEAR, -1);
    break;
  case manBuildingFar:
    soundPlaySound(lpDSManBuildingFar, IDW_MAN_BUILDING_FAR, -1);
    break;
  case manDyingNear:
    soundPlaySound(lpDSManDyingNear, IDW_MAN_DYING_NEAR, -1);
    break;
  case manDyingFar:
    soundPlaySound(lpDSManDyingFar, IDW_MAN_DYING_FAR, -1);
    break;
  case manLayingMineNear:
    soundPlaySound(lpDSManLayingMineNear, IDW_MAN_LAYING_MINE_NEAR, -1);
    break;
  case mineExplosionNear:
    soundPlaySound(lpDSMineExplosionNear, IDW_MINE_EXPLOSION_NEAR, -1);
    break;
  case mineExplosionFar:
    soundPlaySound(lpDSMineExplosionFar, IDW_MINE_EXPLOSION_FAR, -1);
    break;
  default:
    /* shootFar */
    soundPlaySound(lpDSShootFar, IDW_SHOOTING_FAR, -1);
    break;
  }
}

/*********************************************************
*NAME:          soundIsPlayable
*AUTHOR:        John Morrison
*CREATION DATE: 13/6/00
*LAST MODIFIED: 13/6/00
*PURPOSE:
*  Returns whether the sound system is enabled or not. By
*  enabled I mean an error hasn't stopped us from starting
*  it
*
*ARGUMENTS:
*
*********************************************************/
bool soundIsPlayable() {
  return isPlayable;
}
